[Notifier] Change notifier recipient handling

jschaedl opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tickets       | Fix #35558 <!-- prefix each issue number with "Fix #", if any -->
| License       | MIT
| Doc PR        | tbd.

According to @wouterj's reasoning in #35558 I did the following to merge the `AdminRecipient` with the `Recipient` class:

- introduced a `EmailRecipientInterface` with `getEmail(): string` methods
- introduced a `RecipientInterface` that is extended by `EmailRecipientInterface` and `SmsRecipientInterface`
- changed `Recipient` class which now holds an email and a phone number property and implements the `EmailRecipientInterface` and the `SmsRecipientInterface`
- changed `NoRecipient` class which now implements the `RecipientInterface`
- removed the `AdminRecipient` and fixed all related type-hints
- introduced an `EmailRecipient` and `SmsRecipient`
- changed the type-hint of the `$recipient` argument in `NotifierInterface::send()`, `Notifier::getChannels()`, `ChannelInterface::notifiy()` and `ChannelInterface::supports()` to `RecipientInterface`
- changed the type hint of the `$recipient` argument in the `as*Message()` of the  `EmailNotificationInterface` and `SmsNotificationInterface` which now uses the introduced interfaces
- changed the type hint of the `$recipient` argument in the static `fromNotification()` factory methods in `EmailMessage` and `SmsMessage` which now uses the introduced interfaces 
- changed `EmailChannel` to only support recipients which implement the `EmailRecipientInterface`
- changed `SmsChannel` only supports recipients which implement the `SmsRecipientInterface`

jschaedl committed:
[Notifier] Change notifier recipient handling

wouterj reviewed (approved):
Looks great!

I think we should mention the changed things (the added methods + interfaces and the removed `AdminRecipient` class) in the UPGRADE guide. Experimental means we are allowed to break stuff, but we the breaks have to be documented afaik.

OskarStark commented on src/Symfony/Component/Notifier/Recipient/Recipient.php:
```diff
@@ -35,8 +58,25 @@ public function email(string $email): self
+     *
+     * @return $this
+     */
+    public function phone(string $phone): self
```
Does it makes sense adding this method to the interface and the email() method too?

Just asking

jschaedl replied:
I also thought about this. But these methods are actually not used by the current internal implementations (e.g. Channel classes). They are just convenience methods for the default recipient implementation and shouldn't be in the interface for that reason imho.

OskarStark reviewed (commented):
Did you check the current bridge implementations if there are some changes needed?

jschaedl commented:
@OskarStark 

> Did you check the current bridge implementations if there are some changes needed?

Yes I checked the bridges too. All should be good there.

xabbuh commented on src/Symfony/Component/Notifier/CHANGELOG.md:
```diff
@@ -4,11 +4,28 @@ CHANGELOG
+ * The `EmailRecipientInterface` with `hasEmail():bool` and `getEmail():string`
+   methods was introduced.
+ * [BC BREAK]  Added method `hasPhone():bool` to `SmsRecipientInterface`.
+ * Added `email` and and `phone` property to `Recipient`.
```
> Added `$email` and `$phone` properties to `Recipient`.

xabbuh commented on src/Symfony/Component/Notifier/CHANGELOG.md:
```diff
@@ -4,11 +4,28 @@ CHANGELOG
+   methods was introduced.
+ * [BC BREAK]  Added method `hasPhone():bool` to `SmsRecipientInterface`.
+ * Added `email` and and `phone` property to `Recipient`.
+ * [BC BREAK] The `Recipient` now implements `EmailRecipientInterface` and `SmsRecipientInterface`.
```
that's not a BC break if I don't miss anything

xabbuh commented on src/Symfony/Component/Notifier/CHANGELOG.md:
```diff
@@ -4,11 +4,28 @@ CHANGELOG
+ * [BC BREAK]  Added method `hasPhone():bool` to `SmsRecipientInterface`.
+ * Added `email` and and `phone` property to `Recipient`.
+ * [BC BREAK] The `Recipient` now implements `EmailRecipientInterface` and `SmsRecipientInterface`.
+ * [BC BREAK] The `NoRecipient` now implements `EmailRecipientInterface` and `SmsRecipientInterface`.
```
that's not a BC break if I don't miss anything

xabbuh reviewed (commented):
Having `hasEmail()` and `hasPhone()` methods does not look right to me. I'd prefer dedicated classes that only implement `EmailRecipientInterface` and/or `SmsRecipientInterface` when the required data is actually present.

jschaedl commented:
According to @xabbuh comment 

> Having hasEmail() and hasPhone() methods does not look right to me. I'd prefer dedicated classes that only implement EmailRecipientInterface and/or SmsRecipientInterface when the required data is actually present.

I gave it a try and introduced `EmailRecipient` and `SmsRecipient` in addition to the `Recipient` class: https://github.com/symfony/symfony/pull/35773/commits/7df6936dfdc4f019477c80d2b86c4202e9d960b1

I also made sure that the different Recipient objects are always in a valid state to be able to remove the check for a non empty email and phone in the `EmailChannel` and `SmsChannel`.

Some tests were added as well: https://github.com/symfony/symfony/pull/35773/commits/f45f2463db09f3599a8e94f60fc1af530bad58b0

I think this solution could be the way to go. WDYT?

jschaedl commented on src/Symfony/Component/Notifier/CHANGELOG.md:
```diff
@@ -4,11 +4,24 @@ CHANGELOG
+ * The `EmailRecipientInterface` and `SmsRecipientInterface` now extend the `RecipientInterface`.
+ * The `EmailRecipient` and `SmsRecipient` was introduced
+ * [BC BREAK] Changed the type-hint of the `$recipient` argument in `NotifierInterface::send()`, `Notifier::getChannels()`, `ChannelInterface::notifiy()` and `ChannelInterface::supports()` to `RecipientInterface`.
+ * [BC BREAK] Changed the type-hint of the `$recipient` argument in the static `fromNotification()` factory methods in `EmailMessage` and `SmsMessage` to `EmailRecipientInterface` and `SmsRecipientInterface`.
```
You're right.  The `EmailMessage` and `SmsMessage ` are also final and can't be extended anyway.

xabbuh commented on src/Symfony/Component/Notifier/Recipient/Assert.php:
```diff
@@ -0,0 +1,47 @@
+
+    public static function phone(string $phone): void
+    {
+        if (empty($phone)) {
```
`'' === $phone`

xabbuh commented on src/Symfony/Component/Notifier/Recipient/Assert.php:
```diff
@@ -0,0 +1,47 @@
+
+    public static function email(string $email): void
+    {
+        if (empty($email)) {
```
`'' === $email`

xabbuh commented on src/Symfony/Component/Notifier/Recipient/Assert.php:
```diff
@@ -0,0 +1,47 @@
+ *
+ * @experimental in 5.1
+ */
+final class Assert
```
I am not so convinced that we need this class. If we want to reuse the validation logic, maybe we should rather simply add internal methods in the `EmailRecipient` and `SmsRecipient` classes.

jschaedl replied:
Instead of implementing static validation methods in the `EmailRecipient` and `SmsRecipient` classes, we could also introduce dedicated traits to reuse the validation logic. Something like:

```
trait EmailValidationTrait
{
    private static function ensureEmailIsValid(string $email): void
    {
        if ('' === $email) {
            throw new InvalidArgumentException('The email must be set.');
        }

        if (!preg_match('/^[a-zA-Z0-9.!#$%&\'*+\\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)+$/', $email)) {
            throw new InvalidArgumentException('The email is not valid.');
        }
    }
}
```
```
trait PhoneValidationTrait
{
    private static function ensurePhoneIsValid(string $phone): void
    {
        if ('' === $phone) {
            throw new InvalidArgumentException('The phone number must be set (no spaces, international code like in +3312345678).');
        }

        if (!preg_match('/^\+[0-9]*/', $phone)) {
            throw new InvalidArgumentException('The phone number is not valid (no spaces, international code like in +3312345678).');
        }
    }
}
```

```
class Recipient implements EmailRecipientInterface, SmsRecipientInterface
{
    use EmailValidationTrait;
    use PhoneValidationTrait;

    ...

    public function __construct(string $email, string $phone)
    {
        self::ensureEmailIsValid($email);
        self::ensurePhoneIsValid($phone);

        ...

    }

    ...

}

What do you think?

```

jschaedl replied:
@xabbuh WDYT?

xabbuh commented on src/Symfony/Component/Notifier/Recipient/EmailValidationTrait.php:
```diff
@@ -0,0 +1,31 @@
+use Symfony\Component\Notifier\Exception\InvalidArgumentException;
+
+/**
+ * @author Jan Schädlich <jan.schaedlich@sensiolabs.de>
```
I think we should mark it as `@internal`.

xabbuh commented on src/Symfony/Component/Notifier/Recipient/PhoneValidationTrait.php:
```diff
@@ -0,0 +1,31 @@
+use Symfony\Component\Notifier\Exception\InvalidArgumentException;
+
+/**
+ * @author Jan Schädlich <jan.schaedlich@sensiolabs.de>
```
I think we should mark it as `@internal`.

xabbuh commented on src/Symfony/Component/Notifier/Tests/Recipient/RecipientTest.php:
```diff
@@ -0,0 +1,51 @@
+    /**
+     * @dataProvider provideInvalidData
+     */
+    public function testCannotBeConstructedWithInvalidPhone(string $email, string $phone)
```
What about also adding a `testCannotBeConstructedWithInvalidEmail()` test?

jschaedl replied:
The method name is misleading. I will change it to `testCannotBeConstructedWithInvalidEmailOrPhone()`. The used dataprovider is already covering both cases.

xabbuh commented on src/Symfony/Component/Notifier/CHANGELOG.md:
```diff
@@ -4,11 +4,23 @@ CHANGELOG
 
+ * The `EmailRecipientInterface` with `getEmail():string` methods was introduced.
+ * Added `email` and and `phone` properties to `Recipient`.
+ * The `Recipient` now implements `EmailRecipientInterface` and `SmsRecipientInterface`.
```
I don't think we need to mention that here.

xabbuh commented on src/Symfony/Component/Notifier/CHANGELOG.md:
```diff
@@ -4,11 +4,23 @@ CHANGELOG
+ * The `EmailRecipientInterface` with `getEmail():string` methods was introduced.
+ * Added `email` and and `phone` properties to `Recipient`.
+ * The `Recipient` now implements `EmailRecipientInterface` and `SmsRecipientInterface`.
+ * The `NoRecipient` now implements `EmailRecipientInterface` and `SmsRecipientInterface`.
```
I don't think we need to mention that here.

xabbuh commented on src/Symfony/Component/Notifier/CHANGELOG.md:
```diff
@@ -4,11 +4,23 @@ CHANGELOG
 5.1.0
 -----
 
+ * The `EmailRecipientInterface` with `getEmail():string` methods was introduced.
```
I think we should shorten the entries about the new interfaces:

```
* Added `EmailRecipientInterface` , `SmsRecipientInterface` and `RecipientInterface`.
```

jschaedl commented:
Hi @fabpot,

can you please have a look on this PR. WDYT?

fabpot commented on src/Symfony/Component/Notifier/Recipient/Recipient.php:
```diff
@@ -13,30 +13,34 @@
+    private $phone;
 
-    public function __construct(string $email = '')
+    public function __construct(string $email, string $phone)
```
Making both email and phone required is not possible to enforce. The current AdminRecipient does not enforce it either. It's perfectly valid to have an admin/recipient without a phone or with only a phone number and no email.

jschaedl replied:
Oh yes that's true. How about introducing a check to ensure that at least one of email or phone is not empty? Basically something like this:

```
    public function __construct(string $email = '', string $phone = '')
    {
        if (empty(trim($email)) && empty(trim($phone))) {
            throw new LogicException(sprintf('"%s" needs an email or a phone but both cannot be empty.', static::class));
        }
```

xabbuh replied:
I am not sure I agree here. Making both attributes optional means that I couldn't rely on the interface being implementing when passing recipients around and thus defeating their purpose.

jschaedl replied:
@xabbuh 

I think having `has*()` methods in the Interfaces would solve this issue. We could either make `$email` and `$phone` required or optional and userland code can still rely on the interfaces.

fabpot commented on src/Symfony/Component/Notifier/Notifier.php:
```diff
@@ -54,30 +53,30 @@ public function send(Notification $notification, Recipient ...$recipients): void
             $error = 'you should either pass channels in the constructor, override its "getChannels()" method';
             if (null === $this->policy) {
-                throw new LogicException(sprintf('%s; %s, or configure a "%s".', $errorPrefix, $error, ChannelPolicy::class));
+                throw new LogicException(sprintf('"%s"; "%s", or configure a "%s".', $errorPrefix, $error, ChannelPolicy::class));
```
This one is a fabbot false positive and should be reverted.

fabpot commented on src/Symfony/Component/Notifier/Notifier.php:
```diff
@@ -54,30 +53,30 @@ public function send(Notification $notification, Recipient ...$recipients): void
             }
             if (!$channels = $this->policy->getChannels($notification->getImportance())) {
-                throw new LogicException(sprintf('%s; the "%s" returns no channels for importance "%s"; %s.', $errorPrefix, ChannelPolicy::class, $notification->getImportance(), $error));
+                throw new LogicException(sprintf('"%s"; the "%s" returns no channels for importance "%s"; "%s".', $errorPrefix, ChannelPolicy::class, $notification->getImportance(), $error));
```
This one is a fabbot false positive and should be reverted.

fabpot commented on src/Symfony/Component/Notifier/Recipient/EmailRecipient.php:
```diff
@@ -0,0 +1,36 @@
+ *
+ * @experimental in 5.1
+ */
+class EmailRecipient implements EmailRecipientInterface
```
If we go down this road, I would prefer to not have an EmailRecipient and SmsRecipient class, but traits instead that you can compose in your own Recipent class.

jschaedl replied:
Oh yes, will be much more usable like this. Will change this.

fabpot commented on src/Symfony/Component/Notifier/Recipient/PhoneValidationTrait.php:
```diff
@@ -0,0 +1,33 @@
+ */
+trait PhoneValidationTrait
+{
+    private static function ensurePhoneIsValid(string $phone): void
```
Not a big fan of having validation of Email/Phone in Notifier. First, we know first hand that validating an email is complex and people will want to override the default behavior (same for phone numbers, we might have a provider that does not support having an international code for instance). I would remove the validation and let the developer do it before creating Recipient instances.

jschaedl replied:
Makes total sense to me. Thanks for the explanation. I will remove the validation.

fabpot reviewed (changes requested):
Thanks for the work. Not sure yet if I like this new direction or not :) To be honest, I used to have SmsRecipientInterface and EmailRecipientInterface, but removed them... for reasons I don't remember (perhaps the issues I had back then do not hold anymore anyway as I went through many different iterations of the component implementation). I can see some nice improvements though. And removing the AdminRecipent class seems to be a good idea.

I've just done a first round of comments. I think we need to discuss those before going further.

jschaedl commented on src/Symfony/Component/Notifier/Recipient/SmsRecipientTrait.php:
```diff
@@ -16,18 +16,9 @@
-
-        $this->phone = $phone;
-    }
+    protected $phone;
```
Not sure if this should be protected. As the Recipient class is not final it might be useful when extending it.

jschaedl commented on src/Symfony/Component/Notifier/Recipient/EmailRecipientTrait.php:
```diff
@@ -16,18 +16,9 @@
-
-        $this->email = $email;
-    }
+    protected $email;
```
Not sure if this should be protected.

fabpot replied:
I don't think this is needed as one can use the getter.

jschaedl commented:
@fabpot Thanks for your review. I hope I addressed all of your comments :-) See: https://github.com/symfony/symfony/pull/35773/commits/9368554cfce5e245e913e31b79885c2d16d95834

fabpot commented on src/Symfony/Component/Notifier/Recipient/Recipient.php:
```diff
@@ -11,18 +11,27 @@
+    use SmsRecipientTrait;
 
-    public function __construct(string $email = '')
+    public function __construct(string $email = '', string $phone = '')
```
When setting the phone to an empty string, it will be passed anyway to SMS transports as it implements the interface. Before, we checked for a non-empty phone, which I think must be kept.

jschaedl replied:
I think keeping the non-empty phone check would mean if I forgot to set a phone number even though I would like to send a `SmsMessage`, the `SmsChannel` would be skipped and I don't get any information why my notification was not sent as sms.

I would suggest a different approach: We could make the creation of `SmsMessage` more strict and throw an `InvalidArgumentException` with a proper exception message in case the `$phone` argument is empty. We could apply the same to `EmailMessage`

jschaedl replied:
@fabpot WDYT?

fabpot replied:
Works for me.

fabpot replied:
I think you forgot to enforce a non-empty phone on SmsMessage (and the same for the other messages), right?

jschaedl replied:
Oh no 🙈 thats true. On it.

jschaedl replied:
Fixed: d11567f8d37d6423facd4f12ccd9736345859fc7

fabpot reviewed (changes requested):
Sorry for the delay.

jschaedl commented on src/Symfony/Component/Notifier/CHANGELOG.md:
```diff
@@ -4,11 +4,24 @@ CHANGELOG
+   `RecipientInterface`.
+ * Changed `EmailChannel` to only support recipients which implement the `EmailRecipientInterface`.
+ * Changed `SmsChannel` to only support recipients which implement the `SmsRecipientInterface`.
+ * Added the Mattermost notifier bridge.
```
Will revert it.

jschaedl commented on src/Symfony/Component/Notifier/Recipient/Recipient.php:
```diff
@@ -35,8 +44,13 @@ public function email(string $email): self
 
-    public function getEmail(): string
+    /**
+     * @return $this
```
Comment added.

jschaedl commented on src/Symfony/Component/Notifier/Recipient/Recipient.php:
```diff
@@ -11,18 +11,27 @@
-    public function __construct(string $email = '')
+    public function __construct(string $email = '', string $phone = '')
     {
+        if (empty(trim($email)) && empty(trim($phone))) {
```
Adjusted: https://github.com/symfony/symfony/pull/35773/commits/1dab3f54e425df3cf6296765f7cda02f6681cbe9

jschaedl commented on src/Symfony/Component/Notifier/CHANGELOG.md:
```diff
@@ -4,11 +4,24 @@ CHANGELOG
 5.1.0
```
Moved: https://github.com/symfony/symfony/pull/35773/commits/1dab3f54e425df3cf6296765f7cda02f6681cbe9

jschaedl commented:
Hi @fabpot thanks for your review :-)

I made the requested changes and rebased the branch on master.

jschaedl commented on src/Symfony/Component/Notifier/composer.json:
```diff
@@ -17,7 +17,8 @@
         "php": ">=7.2.5",
-        "symfony/polyfill-php80": "^1.15"
+        "symfony/polyfill-php80": "^1.15",
+        "psr/log": "~1.0"
```
I recognized this was missing, because the `EmailMessageTest` failed locally while using the Notification class that depends on psr/log.

fabpot commented on src/Symfony/Component/Notifier/Tests/Message/EmailMessageTest.php:
```diff
@@ -0,0 +1,24 @@
+<?php
+
+declare(strict_types=1);
```
Should be removed everywhere, we're using that in Symfony.

fabpot commented:
Thank you @jschaedl.

nicolas-grekas commented:
This PR is missing corresponding entries in UPGRADE-5.2.md
@jschaedl would you mind sending a PR to fix this, please?

jschaedl commented:
@nicolas-grekas here you go: https://github.com/symfony/symfony/pull/38691
